File: 백준/Gold/2573. 빙산/빙산.py
# 2573 빙산
# 동서남북으로 닿아 있는 0의 개수만큼 높이가 줄어든다.
# 동서남북으로 연결되어 있으면 하나의 덩어리
# 빙산이 두 덩어리 이상으로 분리되는 최초의 시간을 리턴
import sys
from collections import deque

input = sys.stdin.readline


def bfs(a, b):
    dx = [-1, 0, 1, 0]
    dy = [0, -1, 0, 1]
    q = deque([[a, b]])
    visited[a][b] = True

    while q:
        x, y = q.popleft()

        for k in range(4):
            nx = x + dx[k]
            ny = y + dy[k]

            if 0 <= nx < n and 0 <= ny < m and not visited[nx][ny]:
                if ice[nx][ny] != 0:
                    visited[nx][ny] = True
                    q.append([nx, ny])
                else:
                    cnt[x][y] += 1

    return 1


n, m = map(int, input().split())
ice = []
for i in range(n):
    ice.append(list(map(int, input().split())))
year = 0

while True:
    answer = []
    visited = [[False] * m for _ in range(n)]
    cnt = [[0] * m for _ in range(n)]

    for i in range(n):
        for j in range(m):
            if ice[i][j] != 0 and not visited[i][j]:
                answer.append(bfs(i, j))

    for i in range(n):
        for j in range(m):
            ice[i][j] -= cnt[i][j]
            if ice[i][j] < 0:
                ice[i][j] = 0

    if len(answer) == 0 or len(answer) >= 2:
        break

    year += 1

print(year) if len(answer) >= 2 else print(0)

File: 백준/Gold/2573. 빙산/빙산.js
const path = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = require('fs').readFileSync(path).toString().trim().split('\n');

function solution(input) {
  const [N, M] = input[0].split(' ').map(Number);
  const map = input.slice(1).map((row) => row.split(' ').map(Number));

  let years = 0;

  while (true) {
    const icebergCount = countIcebergs(map, N, M);

    if (icebergCount === 0) {
      console.log(0);
      return;
    }
    
    if (icebergCount > 1) {
      console.log(years);
      return;
    }

    simulateMelting(map, N, M);
    years++;
  }
}

function countIcebergs(map, N, M) {
  let count = 0;
  const visited = Array.from({ length: N }, () => Array(M).fill(false));

  for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
      if (map[i][j] > 0 && !visited[i][j]) {
        bfs(map, visited, i, j, N, M);
        count++;
      }
    }
  }

  return count;
}

function bfs(map, visited, x, y, N, M) {
  const directions = [[1, 0], [-1, 0], [0, 1], [0, -1]];
  const queue = [[x, y]];
  visited[x][y] = true;

  while (queue.length) {
    const [currX, currY] = queue.shift();

    for (const [dx, dy] of directions) {
      const newX = currX + dx;
      const newY = currY + dy;

      if (newX >= 0 && newX < N && newY >= 0 && newY < M && map[newX][newY] > 0 && !visited[newX][newY]) {
        visited[newX][newY] = true;
        queue.push([newX, newY]);
      }
    }
  }
}

function simulateMelting(map, N, M) {
  const melt = Array.from({ length: N }, () => Array(M).fill(0));
  const directions = [[1, 0], [-1, 0], [0, 1], [0, -1]];

  for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
      if (map[i][j] > 0) {
        for (const [dx, dy] of directions) {
          const newX = i + dx;
          const newY = j + dy;

          if (newX >= 0 && newX < N && newY >= 0 && newY < M && map[newX][newY] === 0) {
            melt[i][j]++;
          }
        }
      }
    }
  }

  for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
      map[i][j] = Math.max(0, map[i][j] - melt[i][j]);
    }
  }
}

solution(input);